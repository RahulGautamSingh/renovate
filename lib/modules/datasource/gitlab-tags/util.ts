// import { GlobalConfig } from '../../../config/global';
import { regEx } from '../../../util/regex';
import { joinUrlParts } from '../../../util/url';

export const defaultRegistryUrl = 'https://gitlab.com';

export function getDepHost(registryUrl: string = defaultRegistryUrl): string {
  return registryUrl.replace(regEx(/\/api\/v4$/), '');
}

export function getDefaultRegistryUrl(): string[] {
  return [defaultRegistryUrl];
  // const { platform, endpoint } = GlobalConfig.get();
  // return platform === 'gitlab' && endpoint ? [endpoint] : [defaultRegistryUrl];
}

export function getSourceUrl(
  packageName: string,
  registryUrl?: string
): string {
  const depHost = getDepHost(registryUrl);
  return joinUrlParts(depHost, packageName);
}
