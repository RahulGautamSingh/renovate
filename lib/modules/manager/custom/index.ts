import { logger } from '../../../logger';
import type { ExtractConfig, PackageFile, Result } from '../types';
import managers from './api';

export function extractPackageFile(
  content: string,
  fileName: string,
  config: ExtractConfig
): Result<PackageFile | null> {
  const { customType = 'regex' } = config;
  if (!managers.has(customType)) {
    logger.warn({ customType }, 'Unknown customManagers type');
    return null;
  }

  const m = managers.get(customType)!;
  return m.extractPackageFile(content, fileName, config);
}
export const defaultConfig = {
  pinDigests: false,
};
export const supportedDatasources = ['*'];
