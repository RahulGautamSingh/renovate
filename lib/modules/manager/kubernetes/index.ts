import type { Category } from '../../../constants';
import { DockerDatasource } from '../../datasource/docker';
import { KubernetesApiDatasource } from '../../datasource/kubernetes-api';

export { extractPackageFile } from './extract';

export const url = 'https://kubernetes.io/docs';
export const categories: Category[] = ['kubernetes'];

export const defaultConfig = {
  filePatterns: [],
};

export const supportedDatasources = [
  DockerDatasource.id,
  KubernetesApiDatasource.id,
];
