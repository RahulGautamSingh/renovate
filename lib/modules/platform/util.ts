import hasha from 'hasha';
import type { PlatformPrOptions } from './types';

export function repoFingerprint(
  repoId: number | string,
  endpoint: string | undefined
): string {
  const input = endpoint ? `${endpoint}::${repoId}` : `${repoId}`;
  const fingerprint = hasha(input);
  return fingerprint;
}

export function getNewBranchName(branchName?: string): string | undefined {
  if (branchName && !branchName.startsWith('refs/heads/')) {
    return `refs/heads/${branchName}`;
  }
  return branchName;
}

export function getDefaultPlatformPrOptions(): PlatformPrOptions {
  return {
    azureAutoApprove: false,
    azureWorkItemId: 0,
    bbUseDefaultReviewers: false,
    gitLabIgnoreApprovals: 0,
  };
}
