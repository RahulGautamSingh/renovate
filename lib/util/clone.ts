import { quickStringify } from './stringify';

export function clone<T>(input: T | null = null): T {
  const sortedinput = quickStringify(input);
  return sortedinput ? JSON.parse(sortedinput) : {};
}
