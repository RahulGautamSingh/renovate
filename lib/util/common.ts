import JSON5 from 'json5';
import {
  BITBUCKET_API_USING_HOST_TYPES,
  GITEA_API_USING_HOST_TYPES,
  GITHUB_API_USING_HOST_TYPES,
  GITLAB_API_USING_HOST_TYPES,
} from '../constants';
import { logger } from '../logger';
import * as hostRules from './host-rules';
import { parseUrl } from './url';

/**
 * Tries to detect the `platform` from a url.
 *
 * @param url the url to detect `platform` from
 * @returns matched `platform` if found, otherwise `null`
 */
export function detectPlatform(
  url: string
): 'azure' | 'bitbucket' | 'gitea' | 'github' | 'gitlab' | null {
  const { hostname } = parseUrl(url) ?? {};
  if (hostname === 'dev.azure.com' || hostname?.endsWith('.visualstudio.com')) {
    return 'azure';
  }
  if (hostname === 'bitbucket.org' || hostname?.includes('bitbucket')) {
    return 'bitbucket';
  }
  if (
    hostname &&
    (['gitea.com', 'codeberg.org'].includes(hostname) ||
      hostname.includes('gitea') ||
      hostname.includes('forgejo'))
  ) {
    return 'gitea';
  }
  if (hostname === 'github.com' || hostname?.includes('github')) {
    return 'github';
  }
  if (hostname === 'gitlab.com' || hostname?.includes('gitlab')) {
    return 'gitlab';
  }

  const hostType = hostRules.hostType({ url });

  if (!hostType) {
    return null;
  }

  if (BITBUCKET_API_USING_HOST_TYPES.includes(hostType)) {
    return 'bitbucket';
  }
  if (GITEA_API_USING_HOST_TYPES.includes(hostType)) {
    return 'gitea';
  }
  if (GITHUB_API_USING_HOST_TYPES.includes(hostType)) {
    return 'github';
  }
  if (GITLAB_API_USING_HOST_TYPES.includes(hostType)) {
    return 'gitlab';
  }

  return null;
}

export function parseJsonWithFallback(content: string): any {
  let parsedJson: any;

  try {
    parsedJson = JSON.parse(content);
  } catch (err) {
try {
    parsedJson = JSON5.parse(content);
    logger.warn(
      'JSON5.parse was used to parse the JSON data. Please check your json file'
    );
 } catch (err) {
      logger.warn('Invalid JSON format')
 }
  }

  return parsedJson;
}
