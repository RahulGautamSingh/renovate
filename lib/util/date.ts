import { DateTime } from 'luxon';

const ONE_MINUTE_MS = 60 * 1000;
const ONE_DAY_MS = 24 * 60 * ONE_MINUTE_MS;

export function getElapsedDays(timestamp: string): number {
  return Math.floor(
    (new Date().getTime() - new Date(timestamp).getTime()) / ONE_DAY_MS
  );
}

export function getElapsedMinutes(date: Date): number {
  return Math.floor((new Date().getTime() - date.getTime()) / ONE_MINUTE_MS);
}

export function getElapsedHours(time: Date | string): number {
  const pastTime =
    typeof time === 'string'
      ? DateTime.fromISO(time)
      : DateTime.fromJSDate(time);
  const diff = DateTime.now().diff(pastTime, 'hours');
  return diff.hours;
}
