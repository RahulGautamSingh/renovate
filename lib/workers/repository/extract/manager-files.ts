import is from '@sindresorhus/is';
import type {
  CustomManager,
  ManagerConfig,
  MatchStringsStrategy,
} from '../../../config/types';
import { logger } from '../../../logger';
import {
  extractAllPackageFiles,
  extractPackageFile,
  get,
} from '../../../modules/manager';
import type {
  CustomExtractConfig,
  ExtractConfig,
  PackageFile,
} from '../../../modules/manager/types';
import { readLocalFile } from '../../../util/fs';

export function getExtractConfig(config: ManagerConfig): ExtractConfig {
  const result: ExtractConfig = {
    npmrc: config.npmrc,
    aliases: config.aliases,
    skipInstalls: config.skipInstalls,
    npmrcMerge: config.npmrcMerge,
  };
  return result;
}

export function getCustomExtractConfig(
  config: ManagerConfig & CustomManager
): CustomExtractConfig {
  const result: CustomExtractConfig = {
    matchStrings: config.matchStrings,
    autoReplaceStringTemplate: config.autoReplaceStringTemplate,
    matchStringsStrategy: config.matchStringsStrategy as MatchStringsStrategy,
    depNameTemplate: config.depNameTemplate,
    packageNameTemplate: config.packageNameTemplate,
    datasourceTemplate: config.datasourceTemplate,
    versioningTemplate: config.versioningTemplate,
    depTypeTemplate: config.depTypeTemplate,
    currentValueTemplate: config.currentValueTemplate,
    currentDigestTemplate: config.currentDigestTemplate,
    extractVersionTemplate: config.extractVersionTemplate,
    registryUrlTemplate: config.registryUrlTemplate,
  };

  return result;
}

export async function getManagerPackageFiles(
  config: ManagerConfig | (ManagerConfig & CustomManager)
): Promise<PackageFile[]> {
  const { enabled, manager, fileList } = config;
  const extractConfig =
    manager === 'regex'
      ? getCustomExtractConfig(config as ManagerConfig & CustomManager)
      : getExtractConfig(config);
  logger.trace(`getPackageFiles(${manager})`);
  if (!enabled) {
    logger.debug(`${manager} is disabled`);
    return [];
  }
  // istanbul ignore else
  if (is.nonEmptyArray(fileList)) {
    logger.debug(
      `Matched ${
        fileList.length
      } file(s) for manager ${manager}: ${fileList.join(', ')}`
    );
  } else {
    return [];
  }
  // Extract package files synchronously if manager requires it
  if (get(manager, 'extractAllPackageFiles')) {
    const allPackageFiles = await extractAllPackageFiles(
      manager,
      extractConfig,
      fileList
    );
    if (allPackageFiles) {
      for (const packageFile of allPackageFiles) {
        for (let index = 0; index < packageFile.deps.length; index += 1) {
          packageFile.deps[index].depIndex = index;
        }
      }
    }
    // eslint-disable-next-line @typescript-eslint/no-unnecessary-type-assertion
    return allPackageFiles!;
  }
  const packageFiles: PackageFile[] = [];
  for (const packageFile of fileList) {
    const content = await readLocalFile(packageFile, 'utf8');
    // istanbul ignore else
    if (content) {
      const res = await extractPackageFile(
        manager,
        content,
        packageFile,
        extractConfig
      );
      if (res) {
        for (let index = 0; index < res.deps.length; index += 1) {
          res.deps[index].depIndex = index;
        }
        packageFiles.push({
          packageFile,
          ...res,
        });
      }
    } else {
      logger.debug({ packageFile }, 'packageFile has no content');
    }
  }
  return packageFiles;
}
