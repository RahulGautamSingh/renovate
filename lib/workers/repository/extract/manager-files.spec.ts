import { fs, getConfig, mocked } from '../../../../test/util';
import type { RenovateConfig } from '../../../config/types';
import * as _html from '../../../modules/manager/html';
import * as _fileMatch from './file-match';
import { getManagerPackageFiles } from './manager-files';

jest.mock('./file-match');
jest.mock('../../../modules/manager/html');
jest.mock('../../../util/fs');

const fileMatch = mocked(_fileMatch);
const html = mocked(_html);

describe('workers/repository/extract/manager-files', () => {
  describe('getManagerPackageFiles()', () => {
    let config: RenovateConfig;

    beforeEach(() => {
      jest.resetAllMocks();
      config = getConfig();
    });

    it('returns empty of manager is disabled', async () => {
      const managerConfig = { manager: 'travis', enabled: false, fileList: [] };
      const res = await getManagerPackageFiles(managerConfig);
      expect(res).toHaveLength(0);
    });

    it('returns empty of manager is not enabled', async () => {
      config.enabledManagers = ['npm'];
      const managerConfig = { manager: 'docker', enabled: true, fileList: [] };
      const res = await getManagerPackageFiles(managerConfig);
      expect(res).toHaveLength(0);
    });

    it('skips files if null content returned', async () => {
      const managerConfig = {
        manager: 'npm',
        fileList: [],
        enabled: true,
      };
      fileMatch.getMatchingFiles.mockReturnValue(['package.json']);
      const res = await getManagerPackageFiles(managerConfig);
      expect(res).toHaveLength(0);
    });

    it('returns files with extractPackageFile', async () => {
      const managerConfig = {
        manager: 'html',
        enabled: true,
        fileList: ['Dockerfile'],
      };
      fileMatch.getMatchingFiles.mockReturnValue(['Dockerfile']);
      fs.readLocalFile.mockResolvedValueOnce('some content');
      html.extractPackageFile = jest.fn(() => ({
        deps: [{}, { replaceString: 'abc' }],
      })) as never;
      const res = await getManagerPackageFiles(managerConfig);
      expect(res).toEqual([
        {
          packageFile: 'Dockerfile',
          deps: [{ depIndex: 0 }, { depIndex: 1, replaceString: 'abc' }],
        },
      ]);
    });

    it('returns files with extractAllPackageFiles', async () => {
      const managerConfig = {
        manager: 'npm',
        enabled: true,
        fileList: ['package.json'],
      };
      fileMatch.getMatchingFiles.mockReturnValue(['package.json']);
      fs.readLocalFile.mockResolvedValueOnce(
        '{"dependencies":{"chalk":"2.0.0"}}'
      );
      const res = await getManagerPackageFiles(managerConfig);
      expect(res).toMatchSnapshot([
        {
          packageFile: 'package.json',
          packageJsonType: 'app',
          deps: [
            {
              currentValue: '2.0.0',
              datasource: 'npm',
              depName: 'chalk',
              depType: 'dependencies',
            },
          ],
        },
      ]);
    });

    it('returns files with regex extractAllPackageFiles', async () => {
      const managerConfig = {
        manager: 'regex',
        enabled: true,
        fileList: ['airbyte.yaml'],
        matchStrings: [
          '# renovate: datasource=(?<datasource>[^\\s]*) depName=(?<depName>[^\\s]+)( versioning=(?<versioning>[^\\s]*))?.*\\n.*: ["]?(?<currentValue>v?[^"\\s]+)["]?\n',
        ],
        versioningTemplate:
          '{{#if versioning}}{{{versioning}}}{{else}}semver{{/if}}',
      };
      fileMatch.getMatchingFiles.mockReturnValue(['airbyte.yaml']);
      fs.readLocalFile.mockResolvedValueOnce(
        '# renovate: datasource=github-releases depName=airbytehq/airbyte versioning=maven\nversion: "0.39.0-alpha"\n'
      );
      const res = await getManagerPackageFiles(managerConfig);
      expect(res).toMatchSnapshot([
        {
          packageFile: 'airbyte.yaml',
          deps: [
            {
              depName: 'airbytehq/airbyte',
              currentValue: '0.39.0-alpha',
              datasource: 'github-releases',
              versioning: 'maven',
              replaceString:
                '# renovate: datasource=github-releases depName=airbytehq/airbyte versioning=maven\n' +
                'version: "0.39.0-alpha"\n',
              depIndex: 0,
            },
          ],
          matchStrings: [
            '# renovate: datasource=(?<datasource>[^\\s]*) depName=(?<depName>[^\\s]+)( versioning=(?<versioning>[^\\s]*))?.*\\n.*: ["]?(?<currentValue>v?[^"\\s]+)["]?\n',
          ],
        },
      ]);
    });
  });
});
