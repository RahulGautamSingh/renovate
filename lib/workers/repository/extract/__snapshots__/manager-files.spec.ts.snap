// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`workers/repository/extract/manager-files getManagerPackageFiles() returns files with extractAllPackageFiles 1`] = `
Array [
  Object {
    "constraints": Object {},
    "deps": Array [
      Object {
        "currentValue": "2.0.0",
        "datasource": "npm",
        "depIndex": 0,
        "depName": "chalk",
        "depType": "dependencies",
        "prettyDepType": "dependency",
      },
    ],
    "lernaClient": undefined,
    "lernaPackages": undefined,
    "managerData": Object {
      "hasPackageManager": false,
      "lernaJsonFile": undefined,
      "yarnZeroInstall": false,
    },
    "npmLock": undefined,
    "npmrc": undefined,
    "packageFile": "package.json",
    "packageFileVersion": undefined,
    "packageJsonName": undefined,
    "packageJsonType": "app",
    "pnpmShrinkwrap": undefined,
    "skipInstalls": undefined,
    "yarnLock": undefined,
    "yarnWorkspacesPackages": undefined,
  },
]
`;

exports[`workers/repository/extract/manager-files getManagerPackageFiles() returns files with regex extractAllPackageFiles 1`] = `
Array [
  Object {
    "deps": Array [
      Object {
        "currentValue": "0.39.0-alpha",
        "datasource": "github-releases",
        "depIndex": 0,
        "depName": "airbytehq/airbyte",
        "replaceString": "# renovate: datasource=github-releases depName=airbytehq/airbyte versioning=maven
version: \\"0.39.0-alpha\\"
",
        "versioning": "maven",
      },
    ],
    "matchStrings": Array [
      "# renovate: datasource=(?<datasource>[^\\\\s]*) depName=(?<depName>[^\\\\s]+)( versioning=(?<versioning>[^\\\\s]*))?.*\\\\n.*: [\\"]?(?<currentValue>v?[^\\"\\\\s]+)[\\"]?
",
    ],
    "packageFile": "airbyte.yaml",
  },
]
`;
